File: bots/node/approach_b_http.mjs
// Approach B — Headless HTTP-direct sniper (Node, faster tier than Python).
//
// Same strategy as the Python version, but Node's event loop + V8 give a lower
// per-request overhead, so the release->hit number is typically tighter.
//
//   * keepAlive Agent so the "fire" socket is pooled & warm before the drop
//   * long-poll /status/longpoll; the instant it resolves, fire GET /buy on the
//     pooled (already-open) socket
//
// Run:  node bots/node/approach_b_http.mjs --trials 30

import http from 'node:http';
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const HOST = process.env.MOCK_HOST || '127.0.0.1';
const PORT = Number(process.env.MOCK_PORT || 3000);
const __dirname = path.dirname(fileURLToPath(import.meta.url));
const RESULTS_DIR = path.join(__dirname, '..', '..', 'benchmark', 'results');

const args = process.argv.slice(2);
const getArg = (k, d) => { const i = args.indexOf(k); return i >= 0 ? args[i + 1] : d; };
const TRIALS = Number(getArg('--trials', 30));
const APPROACH = getArg('--approach', 'http-direct');
const LANG = getArg('--lang', 'node');

// One keepAlive agent => sockets are pooled and reused (pre-warmed).
const agent = new http.Agent({ keepAlive: true, maxSockets: 8, maxFreeSockets: 8 });

function req(pathname, { method = 'GET', body = null } = {}) {
  return new Promise((resolve, reject) => {
    const r = http.request(
      { host: HOST, port: PORT, path: pathname, method, agent,
        headers: body ? { 'Content-Type': 'application/json' } : {} },
      (res) => { let d = ''; res.on('data', (c) => (d += c)); res.on('end', () => resolve(d)); }
    );
    r.on('error', reject);
    if (body) r.write(body);
    r.end();
  });
}

async function arm(trial) {
  await req('/control/arm', { method: 'POST',
    body: JSON.stringify({ approach: APPROACH, trial, baseDelayMs: 200, jitterMs: 300 }) });
}

async function trialOnce(trial) {
  await req('/health');                 // warm a pooled socket
  // Arm FIRST (resets -> ARMED, schedules a hidden release >=200ms out), THEN attach
  // the long-poll, so we never read a stale RELEASED state from the previous trial.
  await arm(trial);
  const detect = req('/status/longpoll');
  await detect;                          // resolves the instant the server releases
  const raw = await req(`/buy?approach=${APPROACH}&trial=${trial}`); // fire on warm pool
  const data = JSON.parse(raw);
  return data.ok ? Number(data.latency_ns) : null;
}

function summarize(name, ms) {
  const xs = [...ms].sort((a, b) => a - b);
  const p = (q) => xs[Math.min(xs.length - 1, Math.floor(q * xs.length))];
  const mean = xs.reduce((a, b) => a + b, 0) / xs.length;
  console.log(`\n${name}  (n=${xs.length})  server release->hit, milliseconds`);
  console.log(`  min ${xs[0].toFixed(3)}   median ${p(0.5).toFixed(3)}   ` +
              `p95 ${p(0.95).toFixed(3)}   max ${xs[xs.length - 1].toFixed(3)}   mean ${mean.toFixed(3)}`);
}

const rows = [];
for (let i = 0; i < TRIALS; i++) {
  const ns = await trialOnce(i);
  if (ns != null) rows.push({ approach: APPROACH, lang: LANG, trial: i, server_latency_ns: ns });
  await new Promise((r) => setTimeout(r, 30));
}

fs.mkdirSync(RESULTS_DIR, { recursive: true });
const out = path.join(RESULTS_DIR, `${APPROACH}_${LANG}.csv`);
fs.writeFileSync(out,
  'approach,lang,trial,server_latency_ns,client_fire_ns\n' +
  rows.map((r) => `${r.approach},${r.lang},${r.trial},${r.server_latency_ns},`).join('\n') + '\n');

summarize(`${APPROACH} [${LANG}]`, rows.map((r) => r.server_latency_ns / 1e6));
console.log(`  wrote ${out}`);

agent.destroy();

File: bots/go/approach_b_http.go
// Approach B — Headless HTTP-direct sniper (Go, compiled tier).
//
// Go's compiled goroutine/net stack gives the lowest per-request overhead of the
// runnable options here — typically a tight release->hit distribution with a very
// low floor. Same strategy as the Python/Node versions:
//
//   * a keep-alive Transport so the "fire" socket is pooled & warm before the drop
//   * long-poll /status/longpoll; the instant it returns, fire GET /buy on the pool
//
// Install Go (not present on this machine yet):  https://go.dev/dl/
// Run:  go run bots/go/approach_b_http.go --trials 30
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"
)

var (
	host = envOr("MOCK_HOST", "127.0.0.1")
	port = envOr("MOCK_PORT", "3000")
)

func envOr(k, d string) string {
	if v := os.Getenv(k); v != "" {
		return v
	}
	return d
}

func base() string { return "http://" + host + ":" + port }

// One client with a keep-alive transport => pooled, pre-warmed sockets.
var client = &http.Client{
	Transport: &http.Transport{
		MaxIdleConns:        16,
		MaxIdleConnsPerHost: 16,
		IdleConnTimeout:     60 * time.Second,
		DisableKeepAlives:   false,
	},
	Timeout: 35 * time.Second,
}

func get(path string) (string, error) {
	resp, err := client.Get(base() + path)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	b, _ := io.ReadAll(resp.Body)
	return string(b), nil
}

func arm(approach string, trial int) error {
	body := fmt.Sprintf(`{"approach":%q,"trial":%d,"baseDelayMs":200,"jitterMs":300}`, approach, trial)
	resp, err := client.Post(base()+"/control/arm", "application/json", strings.NewReader(body))
	if err != nil {
		return err
	}
	io.Copy(io.Discard, resp.Body)
	resp.Body.Close()
	return nil
}

func trialOnce(approach string, trial int) (float64, bool) {
	get("/health") // warm a pooled socket

	// Arm FIRST (resets -> ARMED, schedules a hidden release >=200ms out), THEN attach
	// the long-poll in a goroutine, so we never read a stale RELEASED state.
	arm(approach, trial)
	done := make(chan string, 1)
	go func() { s, _ := get("/status/longpoll"); done <- s }()
	<-done // returns the instant the server releases

	raw, err := get(fmt.Sprintf("/buy?approach=%s&trial=%d", approach, trial))
	if err != nil {
		return 0, false
	}
	var data struct {
		Ok        bool   `json:"ok"`
		LatencyNs string `json:"latency_ns"`
	}
	if json.Unmarshal([]byte(raw), &data) != nil || !data.Ok {
		return 0, false
	}
	var ns float64
	fmt.Sscanf(data.LatencyNs, "%f", &ns)
	return ns / 1e6, true // -> ms
}

func main() {
	trials := flag.Int("trials", 30, "number of trials")
	approach := flag.String("approach", "http-direct", "approach label")
	lang := flag.String("lang", "go", "lang label")
	flag.Parse()

	var ms []float64
	var rows []string
	for i := 0; i < *trials; i++ {
		if v, ok := trialOnce(*approach, i); ok {
			ms = append(ms, v)
			rows = append(rows, fmt.Sprintf("%s,%s,%d,%.0f,", *approach, *lang, i, v*1e6))
		}
		time.Sleep(30 * time.Millisecond)
	}

	dir := filepath.Join("benchmark", "results")
	os.MkdirAll(dir, 0o755)
	out := filepath.Join(dir, fmt.Sprintf("%s_%s.csv", *approach, *lang))
	os.WriteFile(out, []byte("approach,lang,trial,server_latency_ns,client_fire_ns\n"+strings.Join(rows, "\n")+"\n"), 0o644)

	sort.Float64s(ms)
	if len(ms) > 0 {
		p := func(q float64) float64 { return ms[min(len(ms)-1, int(q*float64(len(ms))))] }
		sum := 0.0
		for _, v := range ms {
			sum += v
		}
		fmt.Printf("\n%s [%s]  (n=%d)  server release->hit, milliseconds\n", *approach, *lang, len(ms))
		fmt.Printf("  min %.3f   median %.3f   p95 %.3f   max %.3f   mean %.3f\n",
			ms[0], p(0.5), p(0.95), ms[len(ms)-1], sum/float64(len(ms)))
	}
	fmt.Printf("  wrote %s\n", out)
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
